### ISSUE TITLE
Ignoring parsing errors of parseHeaders in client.go

### ISSUE THREAD

**Issue by prepaser** [valyala/fasthttp#1917]
I have to deal with the wrong response header (`Referrer Policy`, not `Referrer-Policy` maybe a typo?) sent by the server, but fasthttp causes an error, the moment I receive the wrong header field, and everything goes down. To deal with this, there is nothing I can do but change the fasthttp code. Simply deleting the validHeaderFieldByte in parseHeaders works fine. Browsers also generally seem to just ignore those fields.

Could you please provide a way to ignore the wrong header field? I think it would be enough to optionally disable header field validation, or provide a way to modify validHeaderFieldByteTable.

**Comment by erikdubbelboer** [comment]
A pull request would be welcome. Preferably matching net/http in behavior.

**Comment by ksw2000** [comment]
In `net/http`, they [cut the header](https://cs.opensource.google/go/go/+/master:src/net/textproto/reader.go;l=548;drc=c4772d30bfbed6cfbfdf92066990b5c6dc4065bb)  by `:` and convert the header key by calling [canonicalMIMEHeaderKey](https://cs.opensource.google/go/go/+/master:src/net/textproto/reader.go;l=742;drc=fafd4477f3d19f2c11a628e6e407ecf9924309c1). When canonicalizing the header key, if there is space in the key, they just [skip it](https://cs.opensource.google/go/go/+/master:src/net/textproto/reader.go;l=754-760;drc=fafd4477f3d19f2c11a628e6e407ecf9924309c1). See [https://go.dev/issue/34540](https://go.dev/issue/34540).

That is to say, when processing headers, there is no error if the header key contains "spaces" in `net/http`. However, if we change the example, it will still result in an error.



See the example below:

**header with the key include space**

```go
func main() {
	rawRequest := "HTTP/1.1 200 OK\r\nRefer Policy: origin\r\n\r\n"

	fmt.Println("net/http:")
	reader := strings.NewReader(rawRequest)
	res, err := http.ReadResponse(bufio.NewReader(reader), nil)
	if err != nil {
		log.Println(err)
	} else {
		fmt.Printf("%#v\n", res.Header)
	}

	fmt.Println("fasthttp:")
	req2 := fasthttp.AcquireResponse()
	defer fasthttp.ReleaseResponse(req2)

	reader.Seek(0, io.SeekStart)
	err = req2.Read(bufio.NewReader(reader))
	if err != nil {
		log.Println(err)
	} else {
		req2.Header.VisitAll(func(key []byte, value []byte) {
			fmt.Println(string(key), string(value))
		})
	}
}
```

```
net/http:
http.Header{"Refer Policy":[]string{"origin"}}
fasthttp:
2024/12/13 08:57:12 error when reading response headers: invalid header key "Refer policy". Buffer size=41, contents: "HTTP/1.1 200 OK\r\nRefer policy: origin\r\n\r\n"
```

**header with empty key**

```go
func main() {
	rawRequest := "HTTP/1.1 200 OK\r\n: empty key\r\n\r\n"

	fmt.Println("net/http:")
	reader := strings.NewReader(rawRequest)
	res, err := http.ReadResponse(bufio.NewReader(reader), nil)
	if err != nil {
		log.Println(err)
	} else {
		fmt.Printf("%#v\n", res.Header)
	}

	fmt.Println("fasthttp:")
	req2 := fasthttp.AcquireResponse()
	defer fasthttp.ReleaseResponse(req2)

	reader.Seek(0, io.SeekStart)
	err = req2.Read(bufio.NewReader(reader))
	if err != nil {
		log.Println(err)
	} else {
		req2.Header.VisitAll(func(key []byte, value []byte) {
			fmt.Println(string(key), string(value))
		})
	}
}
```

```go
net/http:
2024/12/13 08:58:36 malformed MIME header line: : empty key
fasthttp:
2024/12/13 08:58:36 error when reading response headers: invalid header key "". Buffer size=32, contents: "HTTP/1.1 200 OK\r\n: empty key\r\n\r\n"
```

**Comment by slicingmelon** [comment]
Yes, I can confirm the issue too. Temporarily, I wrote an error handler and allowed it to pass any "invalid header" error(s). Then, I used `fasthttp.ResponseHeader.VisitAll()` to get all HTTP response headers k:v, as shown in the code snippet above.

**Comment by ksw2000** [comment]
Hi, I am currently fixing this issue. Here are a few points to note:

1. `net/http` will accept `header` keys that contain spaces, but it will not canonicalize them.
2. `net/http` applies the same mechanism for `trailer` fields that contain spaces.
3. Unit tests need to be modified.

**Comment by erikdubbelboer** [comment]
It's ok to modify test to match the behaviour of net/http.
